drivers/adc/adc_npcx.c

```diff
@@ -295,12 +353,198 @@ static int adc_npcx_read_async(const struct device *dev,
 }
 #endif /* CONFIG_ADC_ASYNC */
 
+#if IS_ENABLED(CONFIG_ADC_THRESHOLD_IRQ)
+static void adc_npcx_set_repetitive(const struct device *dev, int input_ch, int enable)
+{
+	struct adc_reg *const inst = HAL_INSTANCE(dev);
+	struct adc_npcx_data *const data = DRV_DATA(dev);
+
+	/* Stop ADC conversion */
+	inst->ADCCNF |= BIT(NPCX_ADCCNF_STOP);
+
+	if (enable) {
+		/* Turn on ADC */
+		inst->ADCCNF |= BIT(NPCX_ADCCNF_ADCEN);
+		/* Set ADC conversion code to SW conversion mode */
+		SET_FIELD(inst->ADCCNF, NPCX_ADCCNF_ADCMD_FIELD,
+			  NPCX_ADC_SCAN_CONVERSION_MODE);
+		/* Update number of channel to be converted */
+		inst->ADCCS |= BIT(input_ch);
+		/* Set conversion type to repetitive (runs continuously) */
+		inst->ADCCNF |= BIT(NPCX_ADCCNF_ADCRPTC);
+
+		data->repetitive_enabled |= BIT(input_ch);
+		/* Start conversion */
+		inst->ADCCNF |= BIT(NPCX_ADCCNF_START);
+	} else {
+		inst->ADCCS &= ~BIT(input_ch);
+
+		data->repetitive_enabled &= ~BIT(input_ch);
+		if (!data->repetitive_enabled) {
+			/* Turn off ADC */
+			inst->ADCCNF &= ~BIT(NPCX_ADCCNF_ADCEN);
+			/* Set ADC to one-shot mode */
+			inst->ADCCNF &= ~BIT(NPCX_ADCCNF_ADCRPTC);
+		} else {
+			/* Start conversion again */
+			inst->ADCCNF |= BIT(NPCX_ADCCNF_START);
+		}
+	}
+}
+
+static int adc_npcx_config_threshold_irq(const struct device *dev,
+					 const struct adc_threshold_cfg *threshold_cfg)
+{
+	struct adc_reg *const inst = HAL_INSTANCE(dev);
+	struct adc_npcx_data *const data = DRV_DATA(dev);
+	uint16_t thrval;
+	volatile uint16_t *thr_ctrl_reg;
+
+	if (threshold_cfg->channel_id >= NPCX_ADC_CH_COUNT ||
+	    threshold_cfg->threshold_id >= NPCX_ADC_THRESHOLD_COUNT) {
+		return -EINVAL;
+	}
+
+	if (threshold_cfg->assert_mode <= ADC_THR_ASSERT_MODE_NONE ||
+	    threshold_cfg->assert_mode >= ADC_THR_ASSERT_MODE_MAX) {
+		return -ENOTSUP;
+	}
+
+	if (!threshold_cfg->adc_threshold_cb) {
+		return -EINVAL;
+	}
+
+	if (threshold_cfg->assert_value >= NPCX_ADC_VREF_VOL ||
+	    threshold_cfg->assert_value == 0)
+		return -EINVAL;
+
+	adc_context_lock(&data->ctx, false, NULL);
+
+	/* Select threshold control register */
+	thr_ctrl_reg = &inst->THRCTL1 + threshold_cfg->threshold_id;
```

Comment:
NPCX7/9 uses a different address for THRCTL. This should be changed in the register structure.

Reply:
Hi, Nuvoton provides PR:#41106 fix the ADC register structure & adds threshold detector setting for NPCX7/9.
Please check it. Thanks.